File: Go/games/minesweeper.go
package games

import (
	"fmt"
	"image/color"
	"math/rand"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/inpututil"
	"github.com/hajimehoshi/ebiten/v2/text"
	"github.com/hajimehoshi/ebiten/v2/vector"
)

type MinesweeperGame struct {
	rows         int32
	cols         int32
	mines        int32
	cellSize     int32
	width        int32
	height       int32
	offsetX      int32
	offsetY      int32
	board        [][]int
	revealed     [][]bool
	flagged      [][]bool
	gameOver     bool
	won          bool
	firstClick   bool
	running      bool
	tileHidden   *TileCache
	tileRevealed *TileCache
	tileFlagged  *TileCache
	tileMine     *TileCache
}

func NewMinesweeperGame() *MinesweeperGame {
	g := &MinesweeperGame{
		rows:     16,
		cols:     16,
		mines:    40,
		cellSize: 30,
	}
	g.width = g.cols * g.cellSize
	g.height = g.rows * g.cellSize
	g.offsetX = (ScreenWidth - g.width) / 2
	g.offsetY = (ScreenHeight - g.height) / 2
	g.tileHidden = NewTileCache(float32(g.cellSize-2), float32(g.cellSize-2), 4)
	g.tileRevealed = NewTileCache(float32(g.cellSize-2), float32(g.cellSize-2), 4)
	g.tileFlagged = NewTileCache(float32(g.cellSize-2), float32(g.cellSize-2), 4)
	g.tileMine = NewTileCache(float32(g.cellSize-2), float32(g.cellSize-2), 4)
	g.Reset()
	return g
}

func (g *MinesweeperGame) Reset() {
	g.board = make([][]int, g.rows)
	for i := range g.board {
		g.board[i] = make([]int, g.cols)
	}
	g.revealed = make([][]bool, g.rows)
	for i := range g.revealed {
		g.revealed[i] = make([]bool, g.cols)
	}
	g.flagged = make([][]bool, g.rows)
	for i := range g.flagged {
		g.flagged[i] = make([]bool, g.cols)
	}
	g.gameOver = false
	g.won = false
	g.firstClick = true
	g.running = true
}

func (g *MinesweeperGame) placeMines(safeX, safeY int32) {
	placed := int32(0)
	for placed < g.mines {
		x := int32(rand.Intn(int(g.cols)))
		y := int32(rand.Intn(int(g.rows)))
		if g.board[y][x] == -1 {
			continue
		}
		if x >= safeX-1 && x <= safeX+1 && y >= safeY-1 && y <= safeY+1 {
			continue
		}
		g.board[y][x] = -1
		placed++
	}
	for y := int32(0); y < g.rows; y++ {
		for x := int32(0); x < g.cols; x++ {
			if g.board[y][x] == -1 {
				continue
			}
			count := 0
			for dy := int32(-1); dy <= 1; dy++ {
				for dx := int32(-1); dx <= 1; dx++ {
					nx, ny := x+dx, y+dy
					if nx >= 0 && nx < g.cols && ny >= 0 && ny < g.rows {
						if g.board[ny][nx] == -1 {
							count++
						}
					}
				}
			}
			g.board[y][x] = count
		}
	}
}

func (g *MinesweeperGame) reveal(x, y int32) {
	if x < 0 || x >= g.cols || y < 0 || y >= g.rows {
		return
	}
	if g.revealed[y][x] || g.flagged[y][x] {
		return
	}
	g.revealed[y][x] = true
	if g.board[y][x] == -1 {
		g.gameOver = true
		return
	}
	if g.board[y][x] == 0 {
		for dy := int32(-1); dy <= 1; dy++ {
			for dx := int32(-1); dx <= 1; dx++ {
				g.reveal(x+dx, y+dy)
			}
		}
	}
	revealedCount := int32(0)
	for y2 := int32(0); y2 < g.rows; y2++ {
		for x2 := int32(0); x2 < g.cols; x2++ {
			if g.revealed[y2][x2] {
				revealedCount++
			}
		}
	}
	if revealedCount == g.rows*g.cols-g.mines {
		g.won = true
	}
}

func (g *MinesweeperGame) Update() error {
	if inpututil.IsKeyJustPressed(ebiten.KeyEscape) {
		return ebiten.Termination
	}
	if inpututil.IsKeyJustPressed(ebiten.KeyR) {
		g.Reset()
		return nil
	}

	if g.gameOver || g.won {
		return nil
	}

	mx, my := ebiten.CursorPosition()
	x := (int32(mx) - g.offsetX) / g.cellSize
	y := (int32(my) - g.offsetY) / g.cellSize
	inBounds := x >= 0 && x < g.cols && y >= 0 && y < g.rows

	if inBounds && inpututil.IsMouseButtonJustPressed(ebiten.MouseButtonLeft) {
		if g.firstClick {
			g.placeMines(x, y)
			g.firstClick = false
		}
		if !g.flagged[y][x] {
			g.reveal(x, y)
		}
	}
	if inBounds && inpututil.IsMouseButtonJustPressed(ebiten.MouseButtonRight) {
		if !g.revealed[y][x] {
			g.flagged[y][x] = !g.flagged[y][x]
		}
	}
	return nil
}

func (g *MinesweeperGame) Draw(screen *ebiten.Image) {
	screen.Fill(ColorBg)

	DrawRoundedRect(screen, float32(g.offsetX), float32(g.offsetY),
		float32(g.width), float32(g.height), 12, color.RGBA{30, 30, 42, 255})

	numberColors := []color.RGBA{
		{0, 0, 0, 255},
		{0, 0, 255, 255},
		{0, 150, 0, 255},
		{255, 0, 0, 255},
		{0, 0, 180, 255},
		{150, 0, 0, 255},
		{0, 150, 150, 255},
		{0, 0, 0, 255},
	}

	for y := int32(0); y < g.rows; y++ {
		for x := int32(0); x < g.cols; x++ {
			rx := float32(g.offsetX + x*g.cellSize + 1)
			ry := float32(g.offsetY + y*g.cellSize + 1)

			if g.revealed[y][x] {
				if g.board[y][x] == -1 {
					DrawImageAt(screen, g.tileMine.Get(color.RGBA{200, 50, 50, 255}), rx, ry)
					cx := rx + float32(g.cellSize)/2
					cy := ry + float32(g.cellSize)/2
					vector.StrokeLine(screen, cx-8, cy, cx+8, cy, 2, color.Black, true)
					vector.StrokeLine(screen, cx, cy-8, cx, cy+8, 2, color.Black, true)
				} else {
					DrawImageAt(screen, g.tileRevealed.Get(color.RGBA{190, 190, 200, 255}), rx, ry)
					if g.board[y][x] > 0 && FontSmall != nil {
						text.Draw(screen, fmt.Sprint(g.board[y][x]), FontSmall,
							int(rx)+int(g.cellSize)/2-8,
							int(ry)+int(g.cellSize)/2+10,
							numberColors[g.board[y][x]])
					}
				}
			} else if g.flagged[y][x] {
				DrawImageAt(screen, g.tileFlagged.Get(color.RGBA{210, 210, 60, 255}), rx, ry)
				if FontSmall != nil {
					text.Draw(screen, "F", FontSmall,
						int(rx)+int(g.cellSize)/2-8,
						int(ry)+int(g.cellSize)/2+10,
						color.RGBA{255, 50, 50, 255})
				}
			} else {
				DrawImageAt(screen, g.tileHidden.Get(color.RGBA{80, 80, 105, 255}), rx, ry)
			}
		}
	}

	if g.gameOver || g.won {
		DrawOverlay(screen, int(g.offsetX), int(g.offsetY), int(g.width), int(g.height))
		var msg string
		var clr color.Color = ColorWhite
		if g.gameOver {
			msg = "GAME OVER"
		} else {
			msg = "YOU WIN!"
			clr = color.RGBA{0, 255, 0, 255}
		}
		b := text.BoundString(FontBig, msg)
		text.Draw(screen, msg, FontBig,
			int(g.offsetX)+int(g.width)/2-b.Dx()/2,
			int(g.offsetY)+int(g.height)/2-20, clr)
		msg2 := "Press R to restart"
		b2 := text.BoundString(FontSmall, msg2)
		text.Draw(screen, msg2, FontSmall,
			int(g.offsetX)+int(g.width)/2-b2.Dx()/2,
			int(g.offsetY)+int(g.height)/2+30, ColorWhite)
	}
}

func (g *MinesweeperGame) Layout(outsideWidth, outsideHeight int) (int, int) {
	return ScreenWidth, ScreenHeight
}

File: Rust/src/games/minesweeper.rs
use crate::common::*;
use macroquad::prelude::*;
use ::rand::Rng;

const ROWS: i32 = 16;
const COLS: i32 = 16;
const MINES: i32 = 40;
const CELL: f32 = 30.0;
const W: f32 = COLS as f32 * CELL;
const H: f32 = ROWS as f32 * CELL;

pub struct MinesweeperGame {
    board: Vec<Vec<i32>>,
    revealed: Vec<Vec<bool>>,
    flagged: Vec<Vec<bool>>,
    game_over: bool,
    won: bool,
    first_click: bool,
    offset: (f32, f32),
}

impl MinesweeperGame {
    pub fn new() -> Self {
        let mut g = Self {
            board: vec![vec![0; COLS as usize]; ROWS as usize],
            revealed: vec![vec![false; COLS as usize]; ROWS as usize],
            flagged: vec![vec![false; COLS as usize]; ROWS as usize],
            game_over: false,
            won: false,
            first_click: true,
            offset: centered_offset(W, H),
        };
        g.reset();
        g
    }

    fn reset(&mut self) {
        self.board = vec![vec![0; COLS as usize]; ROWS as usize];
        self.revealed = vec![vec![false; COLS as usize]; ROWS as usize];
        self.flagged = vec![vec![false; COLS as usize]; ROWS as usize];
        self.game_over = false;
        self.won = false;
        self.first_click = true;
    }

    fn place_mines(&mut self, safe_x: i32, safe_y: i32) {
        let mut rng = ::rand::thread_rng();
        let mut placed = 0;
        while placed < MINES {
            let x = rng.gen_range(0..COLS);
            let y = rng.gen_range(0..ROWS);
            if self.board[y as usize][x as usize] == -1 {
                continue;
            }
            if (x - safe_x).abs() <= 1 && (y - safe_y).abs() <= 1 {
                continue;
            }
            self.board[y as usize][x as usize] = -1;
            placed += 1;
        }
        // 计周围雷数
        for y in 0..ROWS {
            for x in 0..COLS {
                if self.board[y as usize][x as usize] == -1 {
                    continue;
                }
                let mut count = 0;
                for dy in -1..=1 {
                    for dx in -1..=1 {
                        let nx = x + dx;
                        let ny = y + dy;
                        if nx >= 0 && nx < COLS && ny >= 0 && ny < ROWS {
                            if self.board[ny as usize][nx as usize] == -1 {
                                count += 1;
                            }
                        }
                    }
                }
                self.board[y as usize][x as usize] = count;
            }
        }
    }

    fn reveal(&mut self, x: i32, y: i32) {
        if x < 0 || x >= COLS || y < 0 || y >= ROWS {
            return;
        }
        if self.revealed[y as usize][x as usize] || self.flagged[y as usize][x as usize] {
            return;
        }
        self.revealed[y as usize][x as usize] = true;

        if self.board[y as usize][x as usize] == -1 {
            self.game_over = true;
            return;
        }

        if self.board[y as usize][x as usize] == 0 {
            for dy in -1..=1 {
                for dx in -1..=1 {
                    self.reveal(x + dx, y + dy);
                }
            }
        }

        let revealed_count: i32 = self
            .revealed
            .iter()
            .map(|row| row.iter().filter(|&&r| r).count() as i32)
            .sum();
        if revealed_count == ROWS * COLS - MINES {
            self.won = true;
        }
    }
}

impl Game for MinesweeperGame {
    fn update(&mut self) -> bool {
        if is_key_pressed(KeyCode::Escape) {
            return false;
        }
        if is_key_pressed(KeyCode::R) {
            self.reset();
        }

        if !self.game_over && !self.won {
            let (mx, my) = mouse_position();
            let (ox, oy) = self.offset;
            let x = ((mx - ox) / CELL).floor() as i32;
            let y = ((my - oy) / CELL).floor() as i32;

            if x >= 0 && x < COLS && y >= 0 && y < ROWS {
                // 左键：翻开
                if is_mouse_button_pressed(MouseButton::Left) {
                    if self.first_click {
                        self.place_mines(x, y);
                        self.first_click = false;
                    }
                    if !self.flagged[y as usize][x as usize] {
                        self.reveal(x, y);
                    }
                }
                // 右键：插旗
                if is_mouse_button_pressed(MouseButton::Right) {
                    if !self.revealed[y as usize][x as usize] {
                        self.flagged[y as usize][x as usize] =
                            !self.flagged[y as usize][x as usize];
                    }
                }
            }
        }
        true
    }

    fn draw(&mut self) {
        clear_background(COLOR_BG);
        let (ox, oy) = self.offset;

        draw_rectangle(ox, oy, W, H, Color::new(0.118, 0.118, 0.165, 1.0));

        for y in 0..ROWS {
            for x in 0..COLS {
                let rx = ox + x as f32 * CELL + 1.0;
                let ry = oy + y as f32 * CELL + 1.0;
                let rw = CELL - 2.0;
                let rh = CELL - 2.0;

                if self.revealed[y as usize][x as usize] {
                    if self.board[y as usize][x as usize] == -1 {
                        draw_rounded_rect(rx, ry, rw, rh, 4.0, Color::new(0.784, 0.196, 0.196, 1.0));
                        draw_circle(rx + CELL / 2.0, ry + CELL / 2.0, 9.0, BLACK);
                    } else {
                        draw_rounded_rect(rx, ry, rw, rh, 4.0, Color::new(0.745, 0.745, 0.784, 1.0));
                        let n = self.board[y as usize][x as usize];
                        if n > 0 {
                            let colors = [
                                BLACK,
                                Color::new(0.0, 0.0, 1.0, 1.0),
                                Color::new(0.0, 0.588, 0.0, 1.0),
                                Color::new(1.0, 0.0, 0.0, 1.0),
                                Color::new(0.0, 0.0, 0.706, 1.0),
                                Color::new(0.588, 0.0, 0.0, 1.0),
                                Color::new(0.0, 0.588, 0.588, 1.0),
                                BLACK,
                            ];
                            let t = n.to_string();
                            let dims = measure_text(&t, None, 22, 1.0);
                            draw_text(
                                &t,
                                rx + (CELL - dims.width) / 2.0,
                                ry + CELL / 2.0 + dims.height / 2.0 - 2.0,
                                22.0,
                                colors[n.min(7) as usize],
                            );
                        }
                    }
                } else {
                    let flagged = self.flagged[y as usize][x as usize];
                    let color = if flagged {
                        Color::new(0.824, 0.824, 0.235, 1.0)
                    } else {
                        Color::new(0.314, 0.314, 0.412, 1.0)
                    };
                    draw_rounded_rect(rx, ry, rw, rh, 4.0, color);
                    if flagged {
                        let dims = measure_text("F", None, 22, 1.0);
                        draw_text(
                            "F",
                            rx + (CELL - dims.width) / 2.0,
                            ry + CELL / 2.0 + dims.height / 2.0 - 2.0,
                            22.0,
                            Color::new(1.0, 0.196, 0.196, 1.0),
                        );
                    }
                }
            }
        }

        if self.game_over {
            draw_overlay(ox, oy, W, H, 0.6, BLACK);
            draw_text_centered("GAME OVER", ox + W / 2.0, oy + H / 2.0 - 20.0, 40, WHITE);
            draw_text_centered("Press R to restart", ox + W / 2.0, oy + H / 2.0 + 30.0, 22, WHITE);
        } else if self.won {
            draw_overlay(ox, oy, W, H, 0.6, BLACK);
            draw_text_centered("YOU WIN!", ox + W / 2.0, oy + H / 2.0 - 20.0, 40, Color::new(0.0, 1.0, 0.0, 1.0));
            draw_text_centered("Press R to restart", ox + W / 2.0, oy + H / 2.0 + 30.0, 22, WHITE);
        }
    }
}